crates/bitwarden/src/platform/fido2/types.rs

```diff
@@ -0,0 +1,256 @@
+#[cfg_attr(feature = "mobile", derive(uniffi::Enum))]
+pub enum UV {
+    Discouraged,
+    Preferred,
+    Required,
+}
+
+impl From<UV> for Verification {
+    fn from(value: UV) -> Self {
+        match value {
+            UV::Discouraged => Verification::Discouraged,
+            UV::Preferred => Verification::Preferred,
+            UV::Required => Verification::Required,
+        }
+    }
+}
+
+impl From<Verification> for UV {
+    fn from(value: Verification) -> Self {
+        match value {
+            Verification::Discouraged => UV::Discouraged,
+            Verification::Preferred => UV::Preferred,
+            Verification::Required => UV::Required,
+        }
+    }
+}
+
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct GetAssertionResult {
+    pub credential_id: Vec<u8>,
+    pub authenticator_data: Vec<u8>,
+    pub signature: Vec<u8>,
+    pub user_handle: Vec<u8>,
+    /**
+     * SDK IMPL NOTE: This is not part of the spec and is not returned by passkey-rs.
+     * The SDK needs to add this after the response from passkey-rs is received.
+     */
+    pub selected_credential: SelectedCredential,
+}
+
+#[cfg_attr(feature = "mobile", derive(uniffi::Enum))]
+pub enum ClientData {
+    DefaultWithExtraData { android_package_name: String },
+    DefaultWithCustomHash { hash: Vec<u8> },
+}
+
+// To keep compatibility with passkey-rs ClientData, we want extra_client_data to return a type
+// that serializes to Object | Unit. If we just used an Option it would serialize to Object | None
+// So we create our own wrapper types that serialize to the correct values and make the ClientData
+// implementation return that.
+#[derive(Clone)]
+pub(super) struct OptionalAndroidClientData {
+    data: Option<AndroidClientData>,
+}
+
+#[derive(Serialize, Clone)]
+struct AndroidClientData {
+    android_package_name: String,
+}
+
+impl Serialize for OptionalAndroidClientData {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match &self.data {
+            Some(d) => d.serialize(serializer),
+            None => serde::Serializer::serialize_unit(serializer),
+        }
+    }
+}
+
+impl passkey::client::ClientData<OptionalAndroidClientData> for ClientData {
+    fn extra_client_data(&self) -> OptionalAndroidClientData {
+        let data = match self {
+            ClientData::DefaultWithExtraData {
+                android_package_name,
+            } => Some(AndroidClientData {
+                android_package_name: android_package_name.clone(),
+            }),
+            ClientData::DefaultWithCustomHash { .. } => None,
+        };
+
+        OptionalAndroidClientData { data }
+    }
+
+    fn client_data_hash(&self) -> Option<Vec<u8>> {
+        match self {
+            ClientData::DefaultWithExtraData { .. } => None,
+            ClientData::DefaultWithCustomHash { hash } => Some(hash.clone()),
+        }
+    }
+}
```

Comment:
thought: yeah this isn't the best, I didn't realize that you'd have to go to such lengths to make the data optional

Reply:
Part of that was me trying to make sure that values got serialized exactly the same (so the return of `extra_client_data` serialized as Unit instead of None when using `DefaultWithCustomHash`).

I've now tested it and serializing as None works just the same, so I've simplified the code a bit. We still need some code duplication because of the requirement to put the Uniffi tags on the structs anyway so I think we're good.